### PR TITLE
migration KSMCore to feature pattern

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

Based on https://github.com/DataDog/datadog-operator/pull/412, This pull request migrate the logic to configure KSM core to the feature factory pattern.

To achieve it. lots of refactoring was needed:
* create new dedicated package for api `const` values that is shared between v1alpha1 and v2alpha1
* create shared structure between v1alpha1 and v2alpha1
* move some functions present in the `controllers/datadogagent` package in dedicated package to avoid circular dependencies

### Motivation

Provide an example about how to structure the features implementations in the future.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: controllers/datadogagent/feature/kubernetesstatecore/feature.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+)
+
+func init() {
+	err := feature.Register(feature.KubernetesStateCoreIDType, buildKSMfeature)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func buildKSMfeature(options *feature.Options) feature.Feature {
```
Comment (celenechang) [inline controllers/datadogagent/feature/kubernetesstatecore/feature.go:33]:
```suggestion
func buildKSMFeature(options *feature.Options) feature.Feature {
```

File: controllers/datadogagent/controller.go [hunk 2]
```diff
@@ -139,11 +142,15 @@ func (r *Reconciler) reconcileInstance(ctx context.Context, logger logr.Logger,
...
 	// Manage dependencies
 	// -----------------------
-	depsStore := dependencies.NewStore(&dependencies.StoreOptions{SupportCilium: r.options.SupportCilium})
-	rbacManager := feature.NewResourcesManagers(depsStore)
+	storeOptions := &dependencies.StoreOptions{
+		SupportCilium: r.options.SupportCilium,
+		Logger:        logger,
+	}
+	depsStore := dependencies.NewStore(storeOptions)
+	resourcesManager := feature.NewResourcesManagers(depsStore)
```
Comment (celenechang) [inline controllers/datadogagent/controller.go:150]: nit for a future PR, i think it'd be more appropriate to call the struct `ResourceManagers` (drop the `s` in `Resources`)

File: controllers/datadogagent/feature/kubernetesstatecore/feature.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		Name:  apicommon.DDIgnoreAutoConf,
+		Value: "kubernetes_state",
+	}
+
+	return managers.EnvVar().AddEnvVarToContainerWithMergeFunc(common.AgentContainerName, ignoreAutoConf, merger.AppendToValueEnvVarMergeFunction)
+}
+
+// ManageClusterCheckRunnerAgent allows a feature to configure the ClusterCheckRunnerAgent's corev1.PodTemplateSpec
+// It should do nothing if the feature doesn't need to configure it.
+func (f *ksmFeature) ManageClusterCheckRunnerAgent(managers feature.PodTemplateManagers) error {
```
Comment (celenechang) [inline controllers/datadogagent/feature/kubernetesstatecore/feature.go:172]: another nit, but so i don't forget (for naming consistency)
```suggestion
func (f *ksmFeature) ManageClusterChecksRunner(managers feature.PodTemplateManagers) error {
```

File: controllers/datadogagent/common/const.go [hunk 4]
```diff
@@ -13,8 +13,11 @@ const (
 	ClusterAgentSuffix = "dca"
 )
 
+// ContainerName Container name type
+type ContainerName string
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply controllers/datadogagent/common/const.go:17]: we can but in that case we need to move it in `apis/datadoghq/common/v1`

**Pull request merged**